### PR TITLE
[docs] Reproducible pipelines

### PR DESCRIPTION
**Pull request by stevhliu**
This PR refactors the "Improve image quality with deterministic generation" and "Create reproducible pipelines" docs together and repositions deterministic batch generation as an example use case of why creating reproducible pipelines is useful.

- adds the `randn_tensor` utility to the docs
- update path in `stable_diffusion.md` to `reusing_seeds.md` when discussing reproducibility

### CODE REVIEW

File: docs/source/en/stable_diffusion.md [hunk 1]
```diff
@@ -49,7 +49,7 @@ One of the simplest ways to speed up inference is to place the pipeline on a GPU
 pipeline = pipeline.to("cuda")
 ```
 
-To make sure you can use the same image and improve on it, use a [`Generator`](https://pytorch.org/docs/stable/generated/torch.Generator.html) and set a seed for [reproducibility](./using-diffusers/reproducibility):
+To make sure you can use the same image and improve on it, use a [`Generator`](https://pytorch.org/docs/stable/generated/torch.Generator.html) and set a seed for [reproducibility](./using-diffusers/reusing_seeds):
```
Comment (sayakpaul) [inline docs/source/en/stable_diffusion.md:52]: I think you have made sure already but double-checking just in case. 

Did we update it everwhere where this is applicable?

File: docs/source/en/using-diffusers/reusing_seeds.md [hunk 2]
```diff
@@ -10,72 +10,187 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
+> [!TIP]
+> It might seem unintuitive to pass `Generator` objects to a pipeline instead of the integer value representing the seed. However, this is the recommended design when working with probabilistic models in PyTorch because a `Generator` is a *random state* that can be passed to multiple pipelines in a sequence.
+
+</hfoption>
+<hfoption id="GPU">
+
+Writing a reproducible pipeline on a GPU is a bit trickier, and full reproducibility across different hardware is not guaranteed because matrix multiplication - which diffusion pipelines require a lot of - is less deterministic on a GPU than a CPU. For example, if you run the same code example from the CPU example, you'll get a different result even though the seed is identical. This is because the GPU uses a different random number generator than the CPU.
 
 ```python
-generator = [torch.Generator(device="cuda").manual_seed(i) for i in range(4)]
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/reusing_seeds.md:41]: Do you think we should keep an example showing that users can pass a list of generators too?

File: docs/source/en/using-diffusers/reusing_seeds.md [hunk 3]
```diff
@@ -10,72 +10,187 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
-For example, if you want to use the same seed to create 4 identical images:
+Finally, more complex pipelines such as [`UnCLIPPipeline`], are often extremely
+susceptible to precision error propagation. Don't expect similar results across
+different GPU hardware or even PyTorch versions. You'll need to use
+exactly the same hardware and PyTorch version for full reproducibility.
+
+</hfoption>
+</hfoptions>
+
+## Deterministic algorithms
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/reusing_seeds.md:107]: Very nice addition.

**Review by sayakpaul — COMMENTED** [review]
Overall, this looks really good to me. 

Just a thought:

I think it might be nice to include a note about the advanced nature of the generator we pass to the pipeline calls. More specifically, their states are changed "in place" the moment they are consumed.

So,

```python
generator = torch.manual_seed(0)

image_1 = pipeline("...", generator=generator)
image_2 = pipeline("...", generator=generator)
image_3 = pipeline("...", generator=torch.manual_seed(0))
```

here, `image_1` and `image_2` won't actually produce the same results but `image_1` and `image_3` will. This is because we're passing the `generator` object directly to produce `image_2` (and its states were already changed. LMK if this is unclear.

File: docs/source/en/stable_diffusion.md (hunk 1 above)
Comment (stevhliu) [reply docs/source/en/stable_diffusion.md:52]: Yeah, also just updated it in the translated docs

File: docs/source/en/using-diffusers/reusing_seeds.md (hunk 2 above)
Comment (stevhliu) [reply docs/source/en/using-diffusers/reusing_seeds.md:41]: Yeah this should be shown on line 174

File: docs/source/en/using-diffusers/reusing_seeds.md [hunk 4]
```diff
@@ -10,72 +10,185 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
+To avoid this issue, Diffusers has a [`~utils.torch_utils.randn_tensor`] function for creating random noise on the CPU, and then moving the tensor to a GPU if necessary. The [`~utils.torch_utils.randn_tensor`] function is used everywhere inside the pipeline, allowing the user to **always** pass a CPU `Generator` even if the pipeline is run on a GPU.
+
+```python
+import torch
+import numpy as np
+from diffusers import DDIMPipeline
+
+ddim = DDIMPipeline.from_pretrained("google/ddpm-cifar10-32", use_safetensors=True)
+ddim.to("cuda")
+generator = torch.manual_seed(0)
```
Comment (yiyixuxu) [inline docs/source/en/using-diffusers/reusing_seeds.md:91]: maybe we should explain that 
`torch.manual_seed()` automatically create a CPU generator

File: docs/source/en/using-diffusers/reusing_seeds.md [hunk 5]
```diff
@@ -10,72 +10,185 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
...
+
+Non-deterministic behavior occurs when operations are launched in more than one CUDA stream. To avoid this, set the environment variable [CUBLAS_WORKSPACE_CONFIG](https://docs.nvidia.com/cuda/cublas/index.html#results-reproducibility) to `:16:8` to only use one buffer size during runtime.
+
+PyTorch typically benchmarks multiple algorithms to select the fastest one, but if you want reproducibility, you should disable this feature because the benchmark may select different algorithms each time. Set [torch.use_deterministic_algorithms](https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html) to `True` to enable deterministic algorithms.
+
+```py
+import os
+import torch
+
+os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":16:8"
```
Comment (yiyixuxu) [inline docs/source/en/using-diffusers/reusing_seeds.md:118]: we have a utility function for this now https://github.com/huggingface/diffusers/blob/142f353e1c638ff1d20bd798402b68f72c1ebbdd/src/diffusers/utils/testing_utils.py#L861

**Review by yiyixuxu — APPROVED** [review]
thanks!
